### ISSUE TITLE
Update operation for Cholesky decomposition

### ISSUE THREAD

**Issue by nestordemeure** [rustsim/nalgebra#652]
The [Wikipedia page](https://en.wikipedia.org/wiki/Cholesky_decomposition#Updating_the_decomposition) for the Cholesky decomposition provides formulas to perform rank one update, rank one downdate and update the decomposition as if rows had been added or removed from the decomposed matrix.

I am interested in the latter operation and found that it is not implemented in [nalgebra](https://www.nalgebra.org/rustdoc/nalgebra/linalg/struct.Cholesky.html).
Would there be interest in a pull request adding those operations (I am unfamiliar with the code base but might have the time to implement it in about a month) ?

**Comment by sebcrozet** [comment]
Hi!

Yes, that would be a great addition! Don't hesitate to ask if you need help implementing this (especially regarding generics; the internals of nalgebra can be quite difficult in that regard).

**Comment by nestordemeure** [comment]
I have implemented `rank_one_update`, the code is [here](https://github.com/nestordemeure/nalgebra/blob/cholupdate/src/linalg/cholesky.rs) and the corresponding test [here](https://github.com/nestordemeure/nalgebra/blob/cholupdate/tests/linalg/cholesky.rs).

All reviews are welcome. I had troubles with the test due to the fact that `$scalar` is *not* a scalar but a `RandScalar<f64>` instead in this test file (it appears to be different depending on the test file, I could change it easily if you are ok with it). Otherwise the code is a close transcription of Eigen's implementation.

I will work on the row addition/deletion tomorow.

**Comment by nestordemeure** [comment]
I found a number of unneeded `StorageMut` in [solve.rs](https://github.com/rustsim/nalgebra/blob/dev/src/linalg/solve.rs), I will include the fix in the Cholesky PR (as I needed the correct trait).

**Comment by nestordemeure** [comment]
I just submited a pull request with all the methods + tests.

I tried my best to follow the patterns of the library but any comment on the code is welcome.

Experience wise my main source of friction was the `DefaultAllocator` trait, I spent a lot of time on seemingly unrelated error message that could, in fact, be solved by using the correct allocator or adding an allocator trait.
I have not found much documentation on the subject but understanding the why and how of the allocator would probably make writing generic code much easier for newcomers.

In the future, if there is some interest, I could implement some linear solvers such as the conjugate gradient.
